### ISSUE TITLE
Optional (nullable) @Nested one-to-one mappings

### ISSUE THREAD

**Issue by xak2000** [jdbi/jdbi#1404]
Is there a way to make `BeanMapper` to map `@Nested` bean to `null` if there is `LEFT JOIN` present in the query, but related (joined) row is absent in the DB?

I found this in [release notes](https://github.com/jdbi/jdbi/blob/master/RELEASE_NOTES)

```
    - Optional mapping of @Nested objects when using BeanMapper, ConstructorMapper, FieldMapper, or
      KotlinMapper. @Nested objects are only mapped when the result set contains columns that match
      the nested object.
```

But what does mean `the result set contains columns that match the nested object`? I think this means that if `LEFT JOIN` with at least one suitable column is present in the query, then nested object will be mapped anyway, even if actually no related row exist in the DB?

This leads to empty nested objects filled with nulls.

For example, there are two beans:

```
class Account {
  Integer id;
  User user;

  @Nested("u")
  void setUser(User user) {
    this.user = user;
  }
}

class User {
  String name;
}
```

DB structure:
```
CREATE TABLE account (
  id INTEGER PRIMARY KEY
)

CREATE TABLE user (
  account_id INTEGER PRIMARY KEY REFERENCES account(id),
  name VARCHAR(100),
)
```

And the query:

```
SELECT a.id, u.name u_name
FROM account a
LEFT JOIN user u ON u.account_id = a.id
WHERE a.id = :accountId
```

So, the `Account` to `User` is optional `one-to-one` relation, where `Account` can have a `User` or not have a `User`.

Using this mappings and query, `BeanMapper` will always create a `User` bean, even if there is no `user` record found in the DB for this `account`. So we end up with `Account[id=123, user=User[name=null]]`.

What I want to achieve in this case is to end up with `Account[id=123, user=null]`.

**Is there are way to achive it?**

If not, maybe it can be a feature-request then. I don't see any reliable way to make sure that related row doesn't exist in the DB when using `LEFT JOIN` like this, but we can add new parameter to `@Nested` annotation, like `presenceColumn`, which sets a column name in result set, that must have `non-null` value to make `BeanMapper` consider the nested bean as "present". For example, it can be `user.account_id` column from example above. If `user.account_id` is null, then assuredly there is no related `User` record exist (because `account_id` is FK).

So, our query becomes:

```
SELECT a.id, u.name u_name, u.account_id u_account_id
FROM account a
LEFT JOIN user u ON u.account_id = a.id
WHERE a.id = :accountId
```

And `@Nested` annotation becomes:

```
  @Nested(value = "u", presenceColumn = 'account_id')
  void setUser(User user) {
    this.user = user;
  }
```

**Comment by TheRealMarnes** [comment]
Isn't this solved by making a User RowMapper yourself that decides to return null if all given values are null? IIRC it's not the job of the BeanMapper for Account to determine how the nested User should be instantiated. It should delegate that work to another factory, which may just be BeanMapper again if no other matches are found. I'm just spitballing off the top of my head here, may be getting it wrong, don't have the manual and code open in other tabs... I might write up a test and investigate the behavior in detail in a few minutes.

**Comment by TheRealMarnes** [comment]
I'm actually having some trouble reproducing your case, could you provide fuller code please?

![image](https://user-images.githubusercontent.com/3950300/51091484-79087080-178b-11e9-8b11-f0de3460d9c9.png)

**Comment by xak2000** [comment]
> Isn't this solved by making a User RowMapper yourself that decides to return null if all given values are null? 

I think it's possible, but the point is to achieve this without writing manual mapper.

> IIRC it's not the job of the BeanMapper for Account to determine how the nested User should be instantiated. It should delegate that work to another factory, which may just be BeanMapper again if no other matches are found.

Interesting. Thank you for clarification! But then my proposal to create new attribute on `@Nested` may be not suitable solution.

Also, I want to clarify, that "if all given values are null" is not really universal solution here, because there can be really related row with all columns (except FK column) are `null`. And at the same time this "FK column" is not part of related bean (`User` bean from my example). So, `BeanMapper` can't decide between these two situations without some hint on "what column to check for nullability" (that must be FK column):

- There is no related row
- There is related row with all columns are null

So, manual bean mapper of course can check any column it wants. But to make it somehow be supported by `BeanMapper` (and analogues) there must be a way to somehow define the column name that will be checked for null.

> I'm actually having some trouble reproducing your case, could you provide fuller code please?

Thank you! I'll try it and post results here.

**Comment by qualidafial** [comment]
`@Nested` returns null for a nested object, if none of the columns in the result set can be matched up to that nested type. For any `@Nested` object, if Jdbi can find a single column that belongs to that class, it will be mapped--regardless of whether the column value is null.

You probably want to use `RowReducer` (see the docs).

**Comment by xak2000** [comment]
@TheRealMarnes I think I know why your test is not failed. You should put `@Nested("u")` on the method itself (not on the argument). When I put it on argument, it is ignored as I see. So it just doesn't map `User` at all, which leads to `null`.

**Comment by TheRealMarnes** [comment]
@xak2000 indeed, thanks! :)

@qualidafial I've never used rowreducers myself yet, is this what you'd have in mind?

```
@Test
public void foo() {
    db.getJdbi().useHandle(h -> {
        Account account = h.createQuery("select 1 as id, null as u_name")
            .mapToBean(Account.class)
            .reduce(new Account(), (a1, a2) -> {
                a1.setUser(a2.getUser().getName() == null ? null : a2.getUser());
                return a1;
            });

        System.out.println(account);
    });
}

@Getter @Setter @ToString
public static class Account {
    private int id;
    private User user;

    @Nested("u")
    public void setUser(User user) {
        this.user = user;
    }
}

@Getter @Setter @ToString
public static class User {
    private String name;
}
```

It works in any case: `Issue1404.Account(id=0, user=null)`

**Comment by qualidafial** [comment]
I'll have to think about that. `@Nested` was originally intended for things from the same database row, not joins. RowReducer will give you much finer control here.

We'd have to come up with some way for Jdbi to know which column(s) to treat as sentinel values if null. Maybe `@NonNull` on the getter/setter/setter param?

**Comment by xak2000** [comment]
> Maybe `@NonNull` on the getter/setter/setter param?

This can be useful strategy, but not always applicable. For instance, when the bean declaration doesn't contain this nullable field. The good example is FK field. It can be used for join but at the same time is not included in the bean declaration. If bean is always represent nested bean, then explicit back reference to outer bean is redundant.

I think about some generic way to implement different "strategies" to determine if nested bean should be mapped.

### Just a thinking:

It can be an annotation, that can be placed as meta-annotation on other annotations, where some field will define strategy implementation class (implementing specific interface with some `boolean decide(ctx)` method). Then some of the concrete annotation can be placed together with `@Nested` annotation and will be read (it's presence will be determined by base meta-annotation) by the same code, that reads `@Nested` annotation.

Something like Java Bean Validation implements custom constraints.

Example:

Meta-annotation (just to be a marker): `@MapNestedIf`:

```
@Target({ ANNOTATION_TYPE })
@Retention(RUNTIME)
@Documented
public @interface MapNestedIf {
    MapNestedDecider decider();
}
```

Concrete annotation:

```
@Target({ FIELD, METHOD, PARAMETER, ANNOTATION_TYPE, TYPE_USE })
@Retention(RUNTIME)
@Documented
@MapNestedIf(decider = ColumnIsNotNullDecider.class)
public @interface MapIfColumnIsNotNull {
    String value(); // Column name
}
```

```
interface MapNestedDecider {
    boolean decide(Annotation annotation, Context ctx); // I'm not sure what this context will be
    // Maybe another arguments to allow decision based on various state
}
```

```
class ColumnIsNotNullDecider implements MapNestedDecider {
    public boolean decide(Annotation annotation, Context ctx) {
        // cast Annotation to MapIfColumnIsNotNull, get it's argument "value" (column name),
        // check if column with "value" is not null in ResultSet and return true/false
    }
}
```

There can be several "standard" implementation of "deciders" out-of-the-box, like "if some column is null" etc, of course. But the strength of this approach is that any user can implement it's own deciders based on required logic.

---

Maybe it is too complex to implement, I don't know. Just an idea to consider. :-)

**Comment by stevenschlansker** [comment]
I'm a little worried about extending `@Nested` in this way.  It looks fine for one-to-one mappings, but it seems like it'd fall apart as soon as you have some kind of many-to-one or many-to-many relationship.

Have you seen `JoinRow` and `JoinRowMapper`?  it allows you to decompose at least part of the join operation: https://github.com/jdbi/jdbi/blob/master/core/src/test/java/org/jdbi/v3/core/mapper/JoinRowMapperTest.java

We'd certainly consider adding new functionality to core to make joins easier, but I'm a little worried that extending `@Nested` in this way isn't quite right.

**Comment by qualidafial** [comment]
This might be premature abstraction. We have a single use case here, let's target that instead of trying to predict every possible corner case.

Off the top of my head I can think of two approaches:

* Add a new attribute to the `Nested` annotation:

```
public class Foo {
    @Nested(value = "prefix",
            nonNullColumn = "foo_id")
    public Bar getBar() { ... }

    ...
}

public class Bar { ... }
```

* Add a new annotation to be used on the property (or properties). If any so-annotated column is null, the mapper will return `null` instead of an object for that row.

```java
public class Foo {
    @Nested("prefix")
    public Bar getBar() { ... }

    ...
}

public class Bar {
    @MapNonNull // strawman name because I can't think of a better one at the moment
    public BarId getId() { ... }
}
```

I lean toward the latter, since that would allow entities to map to null in other scenarios, e.g. when mapping with `JoinRow` or using a `RowReducer`.

**Comment by stevenschlansker** [comment]
If we really just want to have a null guarding column, one possible implementation is simply to wrap the existing mapper in a `MapNullIfColumnIsNull` delegate mapper.

**Comment by xak2000** [comment]
> I'm a little worried about extending `@Nested` in this way. It looks fine for one-to-one mappings, but it seems like it'd fall apart as soon as you have some kind of many-to-one or many-to-many relationship.

Do you mean?

```
@Nested
Set<User> getUsers();
```

If yes, then I don't see how it couldn't be used in this situation. The "decision" to map or not map will just decide to add it to the set or not. So, the empty set will always be created, but each User will be added to the set only when "decider" decides that it is not absent. Maybe the names that I proposed are confusing. I'm not native english speaker, so the names are bad, just bad. :)

> Have you seen `JoinRow` and `JoinRowMapper`?

Yes, I seen them, and really like them. They are more in spirit of DB relationships. But what I don't like with `JoinRowMapper` when use it together with SqlObjects, is that DAO method can't return a final fully mapped bean. It must return `JoinRow` instance, which must be processed then by callee to assemble final object. Or maybe I'm doing something wrong?

> I lean toward the latter, since that would allow entities to map to null in other scenarios, e.g. when mapping with `JoinRow` or using a `RowReducer`.

Isn't this scenario means that to make decision on mapping or not, `Bar` must be mapped first and only then it's `getId()` can be inspected? So it will be mapped in any case, but just ignored then? (unnecessary CPU work)

Also this scenario means that `Bar` must have `getId()` method, which is not always the case.

So, there are situations when only the first approach works.

**Comment by jpmeyer** [comment]
I'm running up against a very similar issue where I would like my nested object to be null if the primary key for the object is null. In my case, I have an mapped object with a nested object and the only field from the query is the primary key int field. I would like the mapping to return a nested object that is null if the id is null, or set the entire object based on the mapper. Rather than annotate the Nested relation, it makes more sense to simply have the mapper always return null if the id is null. I believe this approach will also solve the OPs use case as well.

**Comment by stevenschlansker** [comment]
@xak2000, do you agree that #1543 is a good solution?  It will be included in release 3.9.0.
